### PR TITLE
[IMP] website_slides: update course visibility

### PR DESCRIPTION
**Pull request by nla-odoo**
before this commit,
the wording of the visibility options can be quite confusing.

after this commit,
- Visibility has been updated
	- Visible to all: I see those training to the public
	- Visible to logged users only: I am a B2B business, and I do not want the public to see what I offer + the pricing I have on those
	- Visible to members only: I am a school, I want to invite my students but external visitors should not be able to enroll
- Move the field above the "enroll policy"
- Enroll Policy has been set to invite when visibility is set as a Course attendee  and hide the enroll policy

Task: https://www.odoo.com/web#id=2862003&menu_id=4720&cids=2&action=4043&model=project.task&view_type=form

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/models/slide_channel.py [hunk 1]
```diff
@@ -265,15 +265,18 @@ def _get_default_enroll_msg(self):
...
     enroll_msg = fields.Html(
         'Enroll Message', help="Message explaining the enroll process",
         default=_get_default_enroll_msg, translate=tools.html_translate, sanitize_attributes=False)
     enroll_group_ids = fields.Many2many('res.groups', string='Auto Enroll Groups', help="Members of those groups are automatically added as members of the channel.")
     visibility = fields.Selection([
-        ('public', 'Open To All'), ('members', 'Members Only')],
-        default='public', string='Visibility', required=True,
+        ('everyone', 'Everyone'),
+        ('logged_user', 'Logged Users'),
+        ('course_attendees', 'Course Attendees')
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:278]: The logic behind members (students) hasn't changed either. `members` in eLearning refers to course attendees, and in the current upgrade PR, 'members' are switched to 'course_attendees'. I'm not sure it is necessary.

So, in the same way as above for "members", I would keep the `Selection` key to `"members"` (which can still valuably be **relabeled** as `"Course Attendees"`).

File: addons/website_slides/models/slide_channel.py [hunk 2]
```diff
@@ -304,6 +307,15 @@ def _get_default_enroll_msg(self):
     can_comment = fields.Boolean('Can Comment', compute='_compute_action_rights', compute_sudo=False)
     can_vote = fields.Boolean('Can Vote', compute='_compute_action_rights', compute_sudo=False)
 
+    @api.constrains('enroll')
+    def _check_enroll(self):
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:311]: However, wouldn't a SQL constraint (more efficient) work here?

**Review by flch-odoo — COMMENTED** [review]
Hello @nla-odoo ,
Nicely done! I have a few comments/suggestions targeted at readability/efficiency and size of diff.
Please do not hesitate to let me know if something is not clear or if I was mistaken.

**Review by flch-odoo — COMMENTED** [review]
Hey @nla-odoo 
Here are a few ideas already while I check one final thing. Thanks for your patience.

**Comment by nla-odoo** [comment]
Hello @flch-odoo 

Changes are done.

Thanks

File: addons/website_slides/security/website_slides_security.xml [hunk 3]
```diff
@@ -90,15 +117,62 @@
...
         <record id="rule_slide_slide_not_website" model="ir.rule">
             <field name="name">Slide: public/portal/user: restricted to published or uploaded by user, and either channel member or public channel &amp; (category or previewable)</field>
             <field name="model_id" ref="model_slide_slide"/>
-            <field name="groups" eval="[(4, ref('base.group_public')), (4, ref('base.group_portal')), (4, ref('base.group_user'))]"/>
-            <field name="domain_force">['&amp;',
-    '|',
-        '&amp;', ('channel_id.visibility', '=', 'public'), '|', ('is_category','=', True), ('is_preview', '=', True),
-        ('channel_id.partner_ids', '=', user.partner_id.id),
-    '&amp;', ('channel_id.website_published', '=', True), '|', ('user_id', '=', user.id), ('website_published', '=', True)]</field>
+            <field name="groups" eval="[(4, ref('base.group_public'))]"/>
```
Comment (flch-odoo) [inline addons/website_slides/security/website_slides_security.xml:159]: Also here, if the rule only applies to public users now, can we stick to published & preview/category?

File: addons/website_slides/security/website_slides_security.xml [hunk 4]
```diff
@@ -90,15 +117,62 @@
             <field name="domain_force">[(1, '=', 1)]</field>
         </record>
 
+        <record id="rule_slide_slide_logged_user" model="ir.rule">
```
Comment (flch-odoo) [inline addons/website_slides/security/website_slides_security.xml:120]: Could you push the rule below the "public" one, similarly to how it was ordered for channels just above?

File: addons/website_slides/security/website_slides_security.xml [hunk 5]
```diff
@@ -90,15 +117,62 @@
...
+                            ('website_published', '=', True),
+                    '|',
+                        ('channel_id.partner_ids', '=', user.partner_id.id),
+                        '&amp;',
+                            ('channel_id.visibility', '=', 'public'),
+                            '|',
+                                ('is_category','=', True),
+                                ('is_preview', '=', True),
+                ]
+            </field>
```
Comment (flch-odoo) [inline addons/website_slides/security/website_slides_security.xml:175]: Also here, if the rule only applies to public users now, can we stick to public & published & preview/category?

```suggestion
            <field name="domain_force">[
                '&amp;',
                    '&amp;',
                        ('channel_id.website_published', '=', True),
                        ('website_published', '=', True),
                    '&amp;',
                        ('channel_id.visibility', '=', 'public'),
                        '|',
                            ('is_category','=', True),
                            ('is_preview', '=', True),
                ]
            </field>
```

**Review by flch-odoo — COMMENTED** [review]
Hello @nla-odoo ,
I finally got the opportunity to look into this and tried to simplify the rules a bit with some success it seems.
Could you please check if these changes would actually break something? 
If so, the tests could probably be updated, but if not, then the same changes to rules could probably apply to other models where rules were split between public and connected users.

Please do not hesitate to ping me as necessary :) Have a nice week-end!

**Comment by nla-odoo** [comment]
Hello @flch-odoo 

Changes are done.

Thanks

File: addons/website_slides_forum/security/website_slides_forum_security.xml [hunk 6]
```diff
@@ -3,8 +3,28 @@
...
         <field name="name">Website forum: User can only access to forum related to followed courses</field>
         <field name="model_id" ref="website_forum.model_forum_forum"/>
-        <field name="domain_force">['&amp;', ('slide_channel_ids.website_published', '=', True), '|', ('slide_channel_ids.visibility', '=', 'public'), ('slide_channel_ids.partner_ids', 'in', user.partner_id.id)]</field>
-        <field name="groups" eval="[(4, ref('base.group_public')), (4, ref('base.group_portal')), (4, ref('base.group_user'))]"/>
+        <field name="domain_force">[
+            '&amp;',
+                ('slide_channel_ids.website_published', '=', True),
+                ('slide_channel_ids.visibility', '=', 'public')
+            ]
+        </field>
```
Comment (flch-odoo) [inline addons/website_slides_forum/security/website_slides_forum_security.xml:11]:
```suggestion
        <field name="domain_force">[('slide_channel_ids.website_published', '=', True), ('slide_channel_ids.visibility', '=', 'public')]</field>
```

File: addons/website_slides_forum/security/website_slides_forum_security.xml [hunk 7]
```diff
@@ -16,8 +36,28 @@
...
         <field name="name">Website forum post: User can only access to post linked to forum related to followed courses</field>
         <field name="model_id" ref="website_forum.model_forum_post"/>
-        <field name="domain_force">['&amp;', ('forum_id.slide_channel_ids.website_published', '=', True), '|', ('forum_id.slide_channel_ids.visibility', '=', 'public'), ('forum_id.slide_channel_ids.partner_ids', 'in', user.partner_id.id)]</field>
-        <field name="groups" eval="[(4, ref('base.group_public')), (4, ref('base.group_portal')), (4, ref('base.group_user'))]"/>
+        <field name="domain_force">[
+            '&amp;',
+                ('forum_id.slide_channel_ids.website_published', '=', True),
+                ('forum_id.slide_channel_ids.visibility', '=', 'public')
+            ]
+        </field>
```
Comment (flch-odoo) [inline addons/website_slides_forum/security/website_slides_forum_security.xml:44]:
```suggestion
        <field name="domain_force">[('forum_id.slide_channel_ids.website_published', '=', True), ('forum_id.slide_channel_ids.visibility', '=', 'public')]</field>
```

File: addons/website_slides_forum/security/website_slides_forum_security.xml [hunk 8]
```diff
@@ -27,10 +67,30 @@
...
     <record id="website_slides_forum_public_tag" model="ir.rule">
-        <field name="name">Website slides forum tag: User can only access to tag linked to forum related to followed courses</field>
+        <field name="name">Website slides forum tag: Public User can only access to tag linked to forum related to followed or public courses</field>
+        <field name="model_id" ref="website_forum.model_forum_tag"/>
+        <field name="domain_force">[
+            '&amp;',
+                ('forum_id.slide_channel_ids.website_published', '=', True),
+                ('forum_id.slide_channel_ids.visibility', '=', 'public')
+            ]
+        </field>
```
Comment (flch-odoo) [inline addons/website_slides_forum/security/website_slides_forum_security.xml:77]:
```suggestion
        <field name="domain_force">[('forum_id.slide_channel_ids.website_published', '=', True) ('forum_id.slide_channel_ids.visibility', '=', 'public')]</field>
```

**Review by flch-odoo — COMMENTED** [review]
Hello @spe-odoo ,
Thanks for taking over this Task+PR.
I have only minor comments, I'm sure it'll soon be ready for the next steps.

File: addons/website_sale_slides/data/sale_order_demo.xml [hunk 9]
```diff
@@ -55,6 +55,30 @@
...
+    </record>
+    <record id="sale_order_course_3_line_1" model="sale.order.line">
+        <field name="order_id" ref="sale_order_course_3"/>
+        <field name="name" model="sale.order.line" eval="obj().env.ref('website_sale_slides.product_course_channel_5').get_product_multiline_description_sale()"/>
+        <field name="product_id" ref="website_sale_slides.product_course_channel_5"/>
+        <field name="product_uom_qty">1</field>
+        <field name="product_uom" ref="uom.product_uom_unit"/>
+        <field name="price_unit">200.0</field>
+    </record>
+
```
[the remark this answers is not in the document]
Comment (spe-odoo) [reply addons/website_sale_slides/data/sale_order_demo.xml:81]: Hi @flch-odoo,

After removing the above code (sale order) : 
after_change

Only the 'sales' count will be zero. So Is it ok for you?

File: addons/website_sale_slides/data/product_demo.xml [hunk 10]
```diff
@@ -23,4 +23,16 @@
...
+    <!-- ================================================== -->
+    <record id="product_course_channel_5" model="product.product">
+        <field name="name">Basics of Furniture Creation</field>
+        <field name="standard_price">200.0</field>
+        <field name="list_price">200.0</field>
+        <field name="detailed_type">course</field>
+        <field name="invoice_policy">order</field>
+        <field name="is_published" eval="True"/>
+        <field name="image_1920" type="base64" file="website_sale_slides/static/img/channel_demo_furniture_2.jpg"/>
+    </record>
```
[the remark this answers is not in the document]
Comment (spe-odoo) [reply addons/website_sale_slides/data/product_demo.xml]: @flch-odoo,
Changes are done

File: addons/website_slides/models/slide_channel.py [hunk 11]
```diff
@@ -327,6 +330,18 @@ def _get_default_enroll_msg(self):
     can_comment = fields.Boolean('Can Comment', compute='_compute_action_rights', compute_sudo=False)
     can_vote = fields.Boolean('Can Vote', compute='_compute_action_rights', compute_sudo=False)
 
+    _sql_constraints = [
+        (
+            "check_enroll",
+            "CHECK(visibility != 'members' OR (visibility = 'members' AND enroll = 'invite'))",
```
Comment (std-odoo) [inline addons/website_slides/models/slide_channel.py:336]: can be simplify to
```
"CHECK(visibility != 'members' OR enroll = 'invite')",
```

File: addons/website_slides/security/website_slides_security.xml [hunk 12]
```diff
@@ -29,11 +29,31 @@
...
+            <field name="name">Channel: portal/user: restricted to published, public or connected User</field>
+            <field name="model_id" ref="model_slide_channel"/>
+            <field name="groups" eval="[(4, ref('base.group_portal')), (4, ref('base.group_user'))]"/>
+            <field name="domain_force">[
+                '&amp;',
+                    ('website_published', '=', True),
+                    '|',
+                    '|',
+                        ('visibility', '=', 'public'),
+                        ('visibility', '=', 'connected'),
```
Comment (std-odoo) [inline addons/website_slides/security/website_slides_security.xml:53]: can we simplify it to this ?
```
('visibility', 'in', ('public', 'connected'))
```

File: addons/website_slides/security/website_slides_security.xml [hunk 13]
```diff
@@ -90,15 +110,47 @@
...
+                    '|',
+                        ('user_id', '=', user.id),
+                        '&amp;',
+                            ('website_published', '=', True),
+                            ('channel_id.website_published', '=', True),
+                    '|',
+                        '&amp;',
+                            '|',
+                                ('channel_id.visibility', '=', 'public'),
+                                ('channel_id.visibility', '=', 'connected'),
```
Comment (std-odoo) [inline addons/website_slides/security/website_slides_security.xml:147]: same here

File: addons/website_slides/tests/test_security.py [hunk 14]
```diff
@@ -152,6 +152,52 @@ def test_access_slide_preview(self):
         self.slide.with_user(self.user_portal).read(['name'])
         self.slide.with_user(self.user_public).read(['name'])
 
+    @mute_logger('odoo.models', 'odoo.addons.base.models.ir_rule')
+    def test_access_channel_visibility_public(self):
```
Comment (std-odoo) [inline addons/website_slides/tests/test_security.py:156]: can we add more tests with website_published set to False ? (you can just check the read access)

File: addons/website_slides/tests/test_security.py [hunk 15]
```diff
@@ -152,6 +152,52 @@ def test_access_slide_preview(self):
...
+        self.channel.with_user(self.user_public).read(['name'])
+
+        self.slide.with_user(self.user_officer).read(['name'])
+        self.slide.with_user(self.user_manager).read(['name'])
+        self.slide.with_user(self.user_emp).read(['name'])
+        self.slide.with_user(self.user_portal).read(['name'])
+        self.slide.with_user(self.user_public).read(['name'])
+
+    @mute_logger('odoo.models', 'odoo.addons.base.models.ir_rule')
+    def test_access_channel_visibility_members(self):
```
Comment (std-odoo) [inline addons/website_slides/tests/test_security.py:174]: same here about `website_published`

File: addons/website_slides/tests/test_security.py [hunk 16]
```diff
@@ -152,6 +152,52 @@ def test_access_slide_preview(self):
...
+    def test_access_channel_visibility_connected(self):
+        self.channel.write({'visibility': 'connected'})
+
+        self.channel.with_user(self.user_officer).read(['name'])
+        self.channel.with_user(self.user_manager).read(['name'])
+        self.channel.with_user(self.user_emp).read(['name'])
+        self.channel.with_user(self.user_portal).read(['name'])
+        with self.assertRaises(AccessError):
+            self.channel.with_user(self.user_public).read(['name'])
+
```
Comment (std-odoo) [inline addons/website_slides/tests/test_security.py:200]: can we add tests for the slide.slide model ?

File: addons/website_slides_forum/security/website_slides_forum_security.xml [hunk 17]
```diff
@@ -1,10 +1,25 @@
...
+    <record id="website_slides_forum_signed_in_user" model="ir.rule">
+        <field name="name">Website forum: Signed In user can only access to forum related to courses</field>
+        <field name="model_id" ref="website_forum.model_forum_forum"/>
+        <field name="domain_force">[
+            '&amp;',
+                ('slide_channel_ids.website_published', '=', True),
+                '|',
+                '|',
+                    ('slide_channel_ids.visibility', '=', 'public'),
+                    ('slide_channel_ids.visibility', '=', 'connected'),
```
Comment (std-odoo) [inline addons/website_slides_forum/security/website_slides_forum_security.xml:18]: same here and below (in public, connected)

**Review by std-odoo — COMMENTED** [review]
@spe-odoo @nla-odoo Hi :) 

I added some comments / propositions (it would be nice to add a bit more tests)

Also can you squash the commits ?

Cheers :)

File: addons/website_sale_slides/data/product_demo.xml [hunk 18]
```diff
@@ -12,6 +12,18 @@
         <field name="image_1920" type="base64" file="website_sale_slides/static/img/channel_demo_tree_1.jpg"/>
     </record>
 
+    <!-- CHANNEL 5: Basics of Furniture Creation -->
```
[the remark this answers is not in the document]
Comment (spe-odoo) [reply addons/website_sale_slides/data/product_demo.xml:15]: Hi @std-odoo,

The commits are squashed and changes have been made. Can you please have a look at it?

Thanks

**Comment by spe-odoo** [comment]
Hi @std-odoo,

The commits are squashed and changes have been made. Can you please have a look at it?

Thanks

**Review by std-odoo — APPROVED** [review]
LGTM :)

**Comment by std-odoo** [comment]
@robodoo r+

**Pull request closed without merging**